org.civicrm.reqangular/test/services/before-hash-query-params.service.spec.js

```diff
@@ -0,0 +1,54 @@
+/* eslint-env amd, jasmine */
+
+(function (CRM) {
+  define([
+    'common/angular',
+    'common/angularMocks',
+    'common/services/before-hash-query-params.service'
+  ], function () {
+    'use strict';
+
+    describe('beforeHashQueryParams', function () {
+      var beforeHashQueryParams, testMap;
+
+      beforeEach(module('common.services'));
+      beforeEach(inject(function (_beforeHashQueryParams_) {
+        beforeHashQueryParams = _beforeHashQueryParams_;
+        testMap = [
+          // url, expected query params
+          ['/civicrm/contact/view', { }],
+          ['/civicrm/contact/view/?reset=1', { reset: '1' }],
+          ['/civicrm/contact/view/?reset=1&cid=4', { reset: '1', cid: '4' }],
+          [
+            '/civicrm/contact/view/?reset=1&cid=4&selectedChild=civitasks#/',
+            { reset: '1', cid: '4', selectedChild: 'civitasks' }
+          ],
+          [
+            '/civicrm/contact/view/?reset=1&cid=4&selectedChild=civitasks#/tasks',
+            { reset: '1', cid: '4', selectedChild: 'civitasks' }
+          ],
+          [
+            '/civicrm/contact/view/?reset=1&cid=4&selectedChild',
+            { reset: '1', cid: '4', selectedChild: null }
+          ],
+          [
+            '/civicrm/contact/view/?reset=1&cid=4&',
+            { reset: '1', cid: '4' }
+          ],
+          [
+            '/civicrm/contact/view/?reset=1&cid=4&foo=',
+            { reset: '1', cid: '4', foo: '' }
+          ]
+        ];
+      }));
+
+      describe('beforeHashQueryParams.parse()', function () {
+        it('parses the given url correctly', function () {
+          testMap.forEach(function (testCase) {
+            expect(beforeHashQueryParams.parse(testCase[0])).toEqual(testCase[1]);
```

Comment:
i'm sorry but i really don't like this approach, although i grant that it has its utility in specific cases. By shoving all the test cases in an array, we really lose the expressiveness that one gets from a test suite well written. A good test suite should read almost like a collection of english sentences, and if we keep doing this we completely lose that and instead we are forced to take a look at a bunch of values in an array

This should be rewritten as something like this
```js
describe('when there is no query string in the url', function () {
  it('returns an empty object', function () {
    // /civicrm/contact/view
  });
});

describe('when is a query string, but after the hash', function () {
  it('ignores the query string and returns an empty object', function () {
    // /civicrm/contact/view#/?foo=bar
  });
});

describe('when there is a query string before the hash', function () {
  it('parses the query string and returns key/value pairs in an object', function () {
    // /civicrm/contact/view?foo=bar&baz=bax
  });
});

describe('when there is a query string both before and after the hash', function () {
  it('parses only the query string before the hash', function () {
    // /civicrm/contact/view?foo=bar#/?baz=bax
  });
});

describe('when there is a query string that trails with an ampersand', function () {
  it('parses the query string ignoring the ampersand', function () {
    // /civicrm/contact/view?foo=bar&
  });
});

describe('when there is a query string that trails with a key without value', function () {
  it('parses the query string and assigns a null value to the key', function () {
    // /civicrm/contact/view?foo=bar&baz
    // /civicrm/contact/view?foo=bar&baz=
  });
});
```

of course it's much more code to write, but it's undeniable that one can simply just read at the `describe` and `it` titles and understand immediately how this service behaves, without having to look at the tests implementation.

Imho the object of a test suite is to read well first and then to be written concisely, not the other way around. Not to mention that as a the test author, if you write the test cases in a "conversational" tone, it makes it easier to think about the logic you are implementing

(cc @igorpavlov @reneolivo @deb1990 )